uniswap-v2-core/contracts/UniswapV2Pair.sol

```diff
@@ -85,20 +118,31 @@ contract UniswapV2Pair is IUniswapV2Pair, UniswapV2ERC20 {
         emit Sync(reserve0, reserve1);
     }
 
-    // if fee is on, mint liquidity equivalent to 1/6th of the growth in sqrt(k)
+    function _calcFee(uint _sqrtNewK, uint _sqrtOldK, uint _platformFee, uint _circulatingShares) internal pure returns (uint _sharesToIssue) {
+        // Assert newK & oldK        < uint112
+        // Assert _circulatingShares < FEE_ACCURACY
+        // Assert _circulatingShares < uint112
+
+        uint256 _scaledGrowth = (_sqrtNewK * ACCURACY) / _sqrtOldK;                         // ASSERT: < UINT256
+        uint256 _scaledMultiplier = ACCURACY - (SQUARED_ACCURACY / _scaledGrowth);          // ASSERT: < UINT128
+        uint256 _scaledTargetOwnership = _scaledMultiplier * _platformFee / FEE_ACCURACY;   // ASSERT: < UINT144 during maths, ends < UINT128
+
+        _sharesToIssue = _scaledTargetOwnership.mul(_circulatingShares) / (ACCURACY.sub(_scaledTargetOwnership)); // ASSER: _scaledTargetOwnership < ACCURACY
+    }
+
     function _mintFee(uint112 _reserve0, uint112 _reserve1) private returns (bool feeOn) {
-        address feeTo = IUniswapV2Factory(factory).feeTo();
-        feeOn = feeTo != address(0);
-        uint _kLast = kLast; // gas savings
+        feeOn = platformFee > 0;
+        uint _kLast = kLast;
+
         if (feeOn) {
-            if (_kLast != 0) {
-                uint rootK = Math.sqrt(uint(_reserve0).mul(_reserve1));
-                uint rootKLast = Math.sqrt(_kLast);
-                if (rootK > rootKLast) {
-                    uint numerator = totalSupply.mul(rootK.sub(rootKLast));
-                    uint denominator = rootK.mul(5).add(rootKLast);
-                    uint liquidity = numerator / denominator;
-                    if (liquidity > 0) _mint(feeTo, liquidity);
+            address feeTo = IUniswapV2Factory(factory).feeTo();
+            uint _sqrtNewK = Math.sqrt(uint(_reserve0).mul(_reserve1));
+            uint _sqrtOldK = Math.sqrt(kLast); // gas savings
+        
+            if (_sqrtOldK != 0){
+                if (_sqrtNewK > _sqrtOldK) {
```

Comment:
No fee on exit?

Reply:
`_mintFee` is called inside `mint` and `burn`, UniswapV2's enter and exit functions. `sqrtNew` sqrtOld` should be taken _BEFORE_ processing the mint and burn and `sqrtOld` should be saved _AFTER_ processing the mint and burn. This means we only measure the fee performance between liquidity additions and removals and we don't charge any fees for inserting or removing liquidity.